src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py

```diff
@@ -0,0 +1,148 @@
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author S. Siso, STFC Daresbury Lab
+
+'''This module contains the FoldConditionalReturnExpressionsTrans. '''
+
+from psyclone.psyir.nodes import (Routine, IfBlock, Return,
+                                  UnaryOperation)
+from psyclone.psyGen import Transformation
+from psyclone.psyir.transformations.transformation_error import \
+        TransformationError
+
+
+class FoldConditionalReturnExpressionsTrans(Transformation):
+    ''' Provides a transformation that folds conditional expressions with only
+    a return statement inside so that the Return statement can disapear
+    (because happens at the end of the Routine). For example, the following
+    code:
+
+    >>> subroutine test(i)
+    >>>    if (i < 5) then
+    >>>        return
+    >>>    endif
+    >>>    if (i > 10) then
+    >>>        return
+    >>>    endif
+    >>>    ! CODE
+    >>>  end subroutine
+
+    will be transformed to:
+
+    >>> subroutine test(i)
+    >>>    if (.not.(i < 5)) then
+    >>>        if (.not.(i > 10)) then
+    >>>            ! CODE
+    >>>        endif
+    >>>    endif
+    >>>  end subroutine
+
+    '''
+
+    def __str__(self):
+        return "Fold all conditional expressions with Return statements."
+
+    @property
+    def name(self):
+        '''Returns the name of this transformation as a string.'''
+        return "FoldConditionalReturnExpressionsTrans"
+
+    def validate(self, node, options=None):
+        '''Ensure that it is valid to apply this transformation to the
+        supplied node.
+
+        :param node: the node to validate.
+        :type node: :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the node is not a Routine.
+        '''
+
+        if not isinstance(node, Routine):
+            raise TransformationError("Error in {0} transformation. "
+                                      "This transformation can only be applied"
+                                      " to Routine nodes.".format(self.name))
+
+    def apply(self, node, options=None):
+        '''Apply this transformation to the supplied node.
+
+        :param node: the node to transform.
+        :type node: :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :returns: 2-tuple of new schedule and memento of transform.
+        :rtype: (:py:class:`psyclone.psyGen.InvokeSchedule`, \
+                 :py:class:`psyclone.undoredo.Memento`)
+        '''
+        routine = node
+        self.validate(routine)
```

Comment:
Please include the options argument to `validate`.

Reply:
Done